### PR TITLE
update nested-pie example with donut

### PR DESCRIPTION
**Pull request by ImportanceOfBeingErnest**
### Update nested_pie example
The PR #10944 attempted to provide a new argument to `pie` whose intention was to implement a donut chart API to the `pie` function. While this is not desireable (see arguments in discussion of that PR), what can be taken from this is that users are not aware of the easy options to produce donut charts in matplotlib. 

This PR changes the existing [nested-pie example](https://matplotlib.org/gallery/pie_and_polar_charts/nested_pie.html#sphx-glr-gallery-pie-and-polar-charts-nested-pie-py) from having several rings, but the inner-most filled to the center, to actually have a hole in them. The concepts of the codes remain; yet the result is more towards applications. It should also be more straight forward to adapt from the new example to a filled version (just leaving out the `width` or `bottom` parameter than it was previously to imagine that this example would also be useful to produce a donut chart.

Further this PR unifies the two examples given such that they produce the same plot through two different approaches. The produced chart then looks like this for both example cases:

image

### Create example for labeling pies

In addition this PR now incorporates a new example showing how to label pies with a legend as well as with annotations.

image

## PR Checklist

- [x] Documentation is sphinx and numpydoc compliant
- [x] PEP8 compliance issues fixed
- [x] Incorporate review
- [x] Get example structure straight. Current solution is not acceptable

### CODE REVIEW

**Comment by ImportanceOfBeingErnest** [comment]
[This would be the new output](https://8594-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/pie_and_polar_charts/nested_pie.html#sphx-glr-gallery-pie-and-polar-charts-nested-pie-py).

File: examples/pie_and_polar_charts/nested_pie.py [hunk 1]
```diff
@@ -17,17 +17,30 @@
...
+cout = cm(np.arange(3)*4)
+cin = cm(np.array([1, 2, 5, 6, 9, 10]))
+
+ax.pie(vals.sum(axis=1), radius=1, colors=cout,
+       wedgeprops=dict(width=size, edgecolor='w'))
+
+ax.pie(vals.flatten(), radius=1-size, colors=cin,
+       wedgeprops=dict(width=size, edgecolor='w'))
+
+ax.set(aspect="equal", title='Pie plot with `ax.pie`')
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/nested_pie.py]: As of #10914 aspect="equal" is not needed anymore.

File: examples/pie_and_polar_charts/nested_pie.py [hunk 2]
```diff
@@ -17,17 +17,30 @@
...
-       colors=plt.rcParams["axes.prop_cycle"].by_key()["color"][:vals.shape[1]])
-ax.pie(vals.sum(axis=1), radius=1)
-ax.set(aspect="equal", title='Pie plot with `ax.pie`')
 
+size = 0.3
+vals = np.array([[60., 32.], [37., 40.], [29., 10.]])
+
+cm = plt.get_cmap("tab20c")
+cout = cm(np.arange(3)*4)
+cin = cm(np.array([1, 2, 5, 6, 9, 10]))
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/nested_pie.py]: Can you please make the variables more telling? i.e. `outer_colors`, `inner_colors`, `cmap`.

File: examples/pie_and_polar_charts/nested_pie.py [hunk 3]
```diff
@@ -36,28 +49,29 @@
...
 
-ax.bar(x=left_outer,
-       width=2 * np.pi / 9, bottom=7, color='C2',
-       linewidth=2, edgecolor='w',
-       height=np.zeros_like(left_outer) + 3)
+ax.bar(x=valsleft.flatten(),
+       width=valsnorm.flatten(), bottom=1-2*size, height=size,
+       color=cin, edgecolor='w', linewidth=1, align="edge")
 
 ax.set(title="Pie plot with `ax.bar` and polar coordinates")
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/nested_pie.py]: Out of curiosity: Do we have general recommendations for `ax.set(prop=...)` vs. `ax.set_prop(...)`? I'm using the latter one for single properties.

**Comment by timhoffm** [comment]
Could you add labels as in your original example and/or a legend, so that this looks more like a real-world usecase. Also, showing how to label your data is an important piece of information that should be conveyed in the example.

The nested donuts are already a more complex example. Maybe we should additionally add a simple donut plot, which could then include part of the labelling information.

File: examples/pie_and_polar_charts/nested_pie.py (hunk 3 above)
Comment (anntzer) [reply examples/pie_and_polar_charts/nested_pie.py]: fwiw I really like `.set(foo=bar)` for everything.

**Comment by ImportanceOfBeingErnest** [comment]
There is the [basic pie example](https://matplotlib.org/gallery/pie_and_polar_charts/pie_features.html#sphx-glr-gallery-pie-and-polar-charts-pie-features-py) and the [pie demo2](https://matplotlib.org/gallery/pie_and_polar_charts/pie_demo2.html#sphx-glr-gallery-pie-and-polar-charts-pie-demo2-py) both showing how to add labels. Are you sure this is needed? 

The main point of this example was and is to show (a) nesting, (b) the two paths to get a pie (through pie and polar bars). I would not clutter it with additional labels or legends. 

One thing we apparently do not have is an example of a pie chart with a legend. I could add that. Should that also show a donut then?

File: examples/pie_and_polar_charts/nested_pie.py (hunk 2 above)
Comment (ImportanceOfBeingErnest) [reply examples/pie_and_polar_charts/nested_pie.py]: Agreed.

File: examples/pie_and_polar_charts/nested_pie.py (hunk 1 above)
Comment (timhoffm) [reply examples/pie_and_polar_charts/nested_pie.py]: Agreed. Can we take the aspect out with an additional PR after backporting?

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+fig, (ax, ax2) = plt.subplots(nrows=2, figsize=(6, 6))
+fig.subplots_adjust(top=0.92, bottom=0.08)
+plt.setp((ax, ax2), aspect="equal")
+
+
+###############################################################################
+# Now it's time for the pie. Starting with a pie recipe, we create the data
+# and a list of labels from it.
+#
+# We can `provide a function to the ``autopct`` argument, which will expand
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: extra backtick before "provide"

**Comment by tacaswell** [comment]
How hard would it be to put the nested pie charts into a function that users could copy-paste as a base for customized helper functions?

**Comment by tacaswell** [comment]
Specifically, functions that look like those described at https://matplotlib.org/tutorials/introductory/usage.html#coding-styles

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+data = [float(x.split()[0]) for x in recipe]
+ingredients = [x.split()[-1] for x in recipe]
+
+def func(pct, allvals):
+    absolute = int(pct/100.*np.sum(allvals))
+    return "{:.1f}%\n({:d} g)".format(pct, absolute)
+
+pie = ax.pie(data, autopct=lambda pct: func(pct, data),
+             textprops=dict(color="w"))
+
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: `wedges, texts, autotexts = ax.pie(...` or on a separate line `wedges, texts, autotexts = pie`

**Comment by ImportanceOfBeingErnest** [comment]
Help! :sos:

I created the new example, which now [looks like this](https://8608-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/pie_and_polar_charts/pie_and_donut_labels.html#sphx-glr-gallery-pie-and-polar-charts-pie-and-donut-labels-py).

[image](https://8608-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/pie_and_polar_charts/pie_and_donut_labels.html#sphx-glr-gallery-pie-and-polar-charts-pie-and-donut-labels-py)

because apparently the figure is created at the point where it ... (I don't know actually); but in any case I want it to be created at the bottom of the page of couse, when it got all its content.

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+# The pie? Wait... it's going to be donut, is it not? 
+# Well, as we see here, the donut is a pie, having a certain ``width`` set to
+# the wedges, which is different from its radius. It's as easy as it gets.
+# This is done via the ``wedgeprops`` argument.
+#
+# Labeling the wedges is done via ``annotate``. We first create some
+# dictionaries of common properties, which we can later pass as keyword
+# argument. We then iterate over all wedges and for each
+#
+# * calcualte the angle of the wedge's center
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: calculate

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+          "60 g butter",
+          "100 ml milk",
+          "1/2 package of yeast"]
+
+data = [225, 90, 50, 60, 100, 5]
+
+donut = ax2.pie(data, radius=1, wedgeprops=dict(width=0.5), startangle=-40)
+
+bbox_props = dict(boxstyle="square,pad=0.3", fc="w", ec="k", lw=0.72)
+kw = dict(xycoords='data',textcoords='data',arrowprops=dict(arrowstyle="-"),
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: space after comma

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+# the wedges, which is different from its radius. It's as easy as it gets.
+# This is done via the ``wedgeprops`` argument.
+#
+# Labeling the wedges is done via ``annotate``. We first create some
+# dictionaries of common properties, which we can later pass as keyword
+# argument. We then iterate over all wedges and for each
+#
+# * calcualte the angle of the wedge's center
+# * from that obtain the coordinates of the point at that angle on the 
+#   circonference
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: circumference

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+# * determine the horizontal alignment of the text, depending on which side
+#   of the circle the point lies.
+# * update the connection style with the obtained angle to have the annotation
+#   arrow point outwards from the donut
+# * finally create the annotation with all the previously determined parameters
```
Comment (timhoffm) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: period after every sentence of the bullet list.

**Comment by ImportanceOfBeingErnest** [comment]
> How hard would it be to put the nested pie charts into a function that users could copy-paste as a base for customized helper functions?

Essentially the nested pie chart example can be copied into a function as it is. However I suppose the user needs to create one such function per given number of data points due to the colors being manually chosen from the colormap. Creating the logic to choose n colors with m shades of that color for a general purpose plotter goes a bit beyond the scope of this, doesn't it?  
Also the example data format fixes two rings. Again, specifying a general purpose format for k rings with nesting is probably a bit too much here. 

Or maybe I misunderstood the question, @tacaswell ?

File: examples/pie_and_polar_charts/pie_and_donut_labels.py (hunk 8 above)
Comment (ImportanceOfBeingErnest) [reply examples/pie_and_polar_charts/pie_and_donut_labels.py]: Sir! Conferences always turn in circles! :laughing:

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+fig, (ax, ax2) = plt.subplots(nrows=2, figsize=(6, 6))
+fig.subplots_adjust(top=0.92, bottom=0.08)
+plt.setp((ax, ax2), aspect="equal")
+
+
+###############################################################################
```
Comment (phobson) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: I think this divider is why the plot isn't rendering fully

File: examples/pie_and_polar_charts/pie_and_donut_labels.py (hunk 7 above)
Comment (phobson) [reply examples/pie_and_polar_charts/pie_and_donut_labels.py]: you can still check for PEP8 with something like `pytest -m pep8 --pep8 path_to_example.py

File: examples/pie_and_polar_charts/pie_and_donut_labels.py (hunk 9 above)
Comment (ImportanceOfBeingErnest) [reply examples/pie_and_polar_charts/pie_and_donut_labels.py]: I'm not a native speaker, but shouldn't sentences have a subject to be sentences? I'll go with a comma instead.

File: examples/pie_and_polar_charts/pie_and_donut_labels.py (hunk 7 above)
Comment (ImportanceOfBeingErnest) [reply examples/pie_and_polar_charts/pie_and_donut_labels.py]: @phobson Thanks for the hint, but I think I have the same problem as [this one](https://github.com/conda/conda/issues/3452#issuecomment-355983878); not sure if that can be fixed without destroying the environment I'm currently working with.

**Comment by ImportanceOfBeingErnest** [comment]
As @phobson suggested, 
>I think this divider is why the plot isn't rendering fully

I removed the divider, but then the comments are real comments to the code, not text within the document. See result [here](https://8611-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/gallery/pie_and_polar_charts/pie_and_donut_labels.html#sphx-glr-gallery-pie-and-polar-charts-pie-and-donut-labels-py). I would rather keep the text as text with links etc. than having it as commented code. 

I now opened [an issue about that at sphinx-gallery](https://github.com/sphinx-gallery/sphinx-gallery/issues/363) to see if anything can be done about it. If not, the example would probably need to be produced in two single figures.

**Comment by ImportanceOfBeingErnest** [comment]
I tried squashing the 6 commits into one; now I ended up with 8 commits?? 

But I think apart from me fighting with git, this is good to go now?

**Comment by jklymak** [comment]
There is a Sphinx gallery command you can put in comments to set the thumbnail figure.  I used it in the constrained layout tutorial if you’d like an example ( on phone so hard to look up directly). 

To fix the git tangle suggest you move your modified file somewhere else, do git reset —hard upstream/master and then copy the file back and commit.  This’ll leave you w one commit against master.  Not such a good method if you have many changes but if just one file it works really well.

**Comment by jklymak** [comment]
Ps I don’t usually merge commit.  I do git rebase -i xxxxxxxx where xxxxxxx is the commit hash of the last commit I did not edit.  That lets me squash all changes against that old commit.

**Comment by ImportanceOfBeingErnest** [comment]
This is crazy. There must be something like `git merge all commits from branch <branchname>` without me typing in random 9 digit numbers, getting spammed by editor windows popping up and constantly running into conflicts and out-of-sync stuff.

**Comment by jklymak** [comment]
You shouldn't really have conflicts (no one else has edited this file). Strongly suggest you do what I origially suggested first and reset your branch to upstream/master and start over if you are getting conflicts: it sounds like giving up, but it really is the easiest way to recover from a mess up.  

Of course you need not type in 9 digit numbers (though cutting and pasting from `git log` isn't *that* hard); you can also do `git rebase -i HEAD~9` or however far back you need to go in history.

**Comment by phobson** [comment]
I do:

```
git checkout donut-doc
git fetch upstream
git rebase upstream/master
git rebase -i HEAD~8 # opens text editor, check your gitconfig
# use the fixup option on all but the first line
# save and exit text editor (for vim that's ":wq"
git push origin donut-doc --force
```

**Comment by ImportanceOfBeingErnest** [comment]
öahm... don't know if that was a good idea?! @phobson 
If you look at the commits, I suddenly seem to have collaborators on this PR.

**Comment by jklymak** [comment]
It’s because you did the merge.  Just copy the files aside and reset hard and you’ll be fine.

**Comment by ImportanceOfBeingErnest** [comment]
I think I finally made it. 

Concerning images it's not that one is better than the other; it's just it would have been nice to have them both as thumbnail. That would require to produce come static image, which is probably not desireable.

**Comment by jklymak** [comment]
Great - Git can be a bit of a PITA if you hit a glitch in your workflow.  Thanks for sticking with it. 

I’ll merge once CI passes.  These are great examples!  Thanks...

File: examples/pie_and_polar_charts/pie_and_donut_labels.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+bbox_props = dict(boxstyle="square,pad=0.3", fc="w", ec="k", lw=0.72)
+kw = dict(xycoords='data', textcoords='data', arrowprops=dict(arrowstyle="-"),
+          bbox=bbox_props, zorder=0, va="center")
+
+for i, p in enumerate(wedges):
+    ang = (p.theta2 - p.theta1)/2. + p.theta1
+    y = np.sin(np.deg2rad(ang))
+    x = np.cos(np.deg2rad(ang))
+    horizontalalignment = ["", "left", "right"][int(np.sign(x))]
```
Comment (anntzer) [inline examples/pie_and_polar_charts/pie_and_donut_labels.py]: very minor nitpick: I think this reads better as e.g. `{-1: "right", 1: "right"}[np.sign(x)]`.  Can be done in a separate PR if you prefer this to go in first, I agree it's a very nice example overall.

Comment (ImportanceOfBeingErnest) [reply examples/pie_and_polar_charts/pie_and_donut_labels.py]: I totally agree; that is not very understandable. 

If it's ok with you I would then just do a second commit here without trying to squash merge and unwillingly inviting some other people's PRs to the party. (Still not sure what I did wrong for that to happen.)

**Pull request merged**